crates/lox-io/src/ndm/kvn/parser.rs

```diff
@@ -0,0 +1,1013 @@
+/*
+ * Copyright (c) 2023. Helge Eichhorn and the LOX contributors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+// KVN spec section 7.4 of https://public.ccsds.org/Pubs/502x0b3e1.pdf
+
+use nom::bytes::complete as nb;
+use nom::character::complete as nc;
+use nom::error::{ErrorKind, ParseError};
+use nom::sequence as ns;
+use regex::Regex;
+
+use super::deserializer::KvnDeserializerErr;
+
+#[derive(Debug, PartialEq)]
+pub enum KvnStringParserErr<I> {
+    EmptyKeyword { input: I },
+    EmptyValue { input: I },
+    InvalidFormat { input: I },
+    ParserError(I, ErrorKind),
+}
+
+#[derive(Debug, PartialEq)]
+pub enum KvnKeyMatchErr<I> {
+    KeywordNotFound { expected: I },
+}
+
+#[derive(PartialEq, Debug)]
+pub enum KvnNumberParserErr<I> {
+    ParserError(I, ErrorKind),
+    EmptyKeyword { input: I },
+    EmptyValue { input: I },
+    InvalidFormat { input: I },
+}
+
+#[derive(PartialEq, Debug)]
+pub enum KvnDateTimeParserErr<I> {
+    ParserError(I, ErrorKind),
+    EmptyKeyword { input: I },
+    EmptyValue { input: I },
+    InvalidFormat { input: I },
+}
+
+impl<I> From<nom::Err<KvnStringParserErr<I>>> for KvnDeserializerErr<I> {
+    fn from(value: nom::Err<KvnStringParserErr<I>>) -> Self {
+        match value {
+            nom::Err::Error(KvnStringParserErr::EmptyValue { input })
+            | nom::Err::Failure(KvnStringParserErr::EmptyValue { input }) => {
+                KvnDeserializerErr::EmptyValue { input }
```

Comment:
A lot of this error handling isn't covered by tests, but this mapping is quite important. Are there pathological inputs you can create that exercise them?

Reply:
Is it okay with you if I add a ticket to follow-up on this?

Reply:
Sure thing!